src/Components/Shared/src/DynamicRootComponentInterop.cs

```diff
@@ -0,0 +1,181 @@
+    /// code.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public class DynamicRootComponentInterop : IDisposable
+    {
+        // TODO: Clear on hot reload
+        private static readonly ConcurrentDictionary<Type, ParameterTypeCache> ParameterTypeCaches = new();
+
+        private const int MaxParameters = 100;
+        private readonly DotNetObjectReference<DynamicRootComponentInterop> _selfReference;
+        private readonly IJSRuntime _jsRuntime;
+        private readonly JsonSerializerOptions _jsonOptions;
+        private readonly Func<int, Type> _getRootComponentType;
+        private readonly Func<Type, string, int> _addRootComponent;
+        private readonly Func<int, ParameterView, Task> _renderRootComponentAsync;
+        private readonly Action<int> _removeRootComponent;
+        private readonly Dictionary<string, Type> _allowedComponentTypes;
+
+        internal DynamicRootComponentInterop(
+            DefaultDynamicRootComponentConfiguration configurationBuilder,
+            IJSRuntime jsRuntime,
+            Func<int, Type> getRootComponentType,
+            Func<Type, string, int> addRootComponent,
+            Func<int, ParameterView, Task> renderRootComponentAsync,
+            Action<int> removeRootComponent)
+        {
+            _selfReference = DotNetObjectReference.Create(this);
+            _jsRuntime = jsRuntime;
+            _jsonOptions = configurationBuilder.JsonOptions;
+            _getRootComponentType = getRootComponentType;
+            _addRootComponent = addRootComponent;
+            _removeRootComponent = removeRootComponent;
+            _renderRootComponentAsync = renderRootComponentAsync;
+
+            // Snapshot the config to ensure it's not mutated later
+            _allowedComponentTypes = new(configurationBuilder.AllowedComponentsByIdentifier, StringComparer.Ordinal);
+        }
+
+        internal async ValueTask InitializeAsync()
+        {
+            if (_allowedComponentTypes.Count > 0)
+            {
+                await _jsRuntime.InvokeVoidAsync(
+                    "Blazor._internal.setDynamicRootComponentManager",
+                    _selfReference);
+            }
+        }
+
+        /// <summary>
+        /// For framework use only.
+        /// </summary>
+        [JSInvokable]
+        public int AddRootComponent(string componentIdentifier, string domElementSelector)
+        {
+            if (!_allowedComponentTypes.TryGetValue(componentIdentifier, out var componentType))
+            {
+                throw new ArgumentException($"There is no registered dynamic root component with identifier '{componentIdentifier}'.");
+            }
+
+            return _addRootComponent(componentType, domElementSelector);
+        }
+
+        /// <summary>
+        /// For framework use only.
+        /// </summary>
+        [JSInvokable]
+        public Task RenderRootComponentAsync(int componentId, int parameterCount, byte[] parametersJsonUtf8)
```

Comment:
I did try using `JsonElement` here but it's pretty bad due to the absence of https://github.com/dotnet/runtime/issues/31274